---
title: "Fig4a_isolates_analysis"
author: "Minghao Chia"
date: "2/16/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load libraries

```{r}
library(tidyverse)
library(ggdendro)
library(here)
library(ggtree)
library(reshape2)
```

Read in data files. Note that isolate D12/HYT (WHB11472) only had 400K read pairs after sequencing and was thus not subject to genome assembly. This leaves 56 assemblies for downstream checkM analysis

```{r}
#load all isolate metadata (58)
all_isolate_metadata <- read_tsv("../metadata/manuscript_isolates_metadata.tsv")

#load checkM results for the 57 assemblies
checkm_eval_dat <- read_tsv("../data/raw/checkM_out/final_checkm_eval_dat")

#load fastANI results for the 57 assemblies
fastANI <- read_tsv("../data/raw/fastANI_out/manuscript_fastANI_results")

#load ANI results after running Mummer
#To get the %ANI, refer to the mummer dist object, column X10.
mummer_dist <- read_tsv("../data/raw/mummer_out/mummer_dist_sub.dat",
                        col_names = FALSE)

mummer_dist$pattern <- paste(mummer_dist$X1, mummer_dist$X2)

#X10 is the %ANIM, gSNP adjusted.  
mean(mummer_dist$X10) #99.49552
sd(mummer_dist$X10) #0.249

```

Assemblies used for further analysis must fulfill the following criteria:
i) >= 95% ANI compared to 95% ANI compared to S. aureus reference genome (strain NCTC 8325) 
ii) >= 95% completeness and <= 5% contamination
 

```{r}

isolates_passQC <- checkm_eval_dat %>% filter(completeness >= 95 & contamination <= 5) %>% filter(bin_id %in% (all_isolate_metadata %>% filter(percent_ANI_vs_ref >= 95) %>% pull(LibraryID)))  #55 genomes left

#final metadata for 52 isolates used in the analysis
final_isolate_metadata <- all_isolate_metadata %>% filter(LibraryID %in% isolates_passQC$bin_id)

```


Do a Wilcoxon rank sum test on the %ANI for isolates from the lesions and/or nares, axila and groin of children vs isolates from the nares, axila and groin of caregivers from the same house (matching houses: 6 for AD group, 5 for control group)

Note that the p-value is < 0.05, showing statistical significance. 


```{r wilcox_test}
AD_child_candidates <- c("A07", 
                         "A08",
                         "A11",
                         "A15",
                         "A17",
                         "A18")

AD_cg_candidates <- c("B07", 
                         "B08",
                         "B11",
                         "B15",
                         "B17",
                         "B18")

ctrl_child_candidates <- c("C02", 
                         "C09",
                         "C12",
                         "C17",
                         "C30")

ctrl_cg_candidates <- c("D02", 
                         "D09",
                        "D12",
                         "D17",
                         "D30")

AD_matching_vector <- vector()

for (i in 1:length(AD_child_candidates)){
  match1 <- AD_child_candidates[i]
  match2 <- AD_cg_candidates[i]
  
  row_number_for_subset <- which(str_detect(string=mummer_dist$pattern, 
                                            pattern=match1) & str_detect(string=mummer_dist$pattern, 
                                            pattern=match2))
  
  AD_matching_vector <-c(AD_matching_vector,row_number_for_subset)
}


matched_AD_mummer_dist <- mummer_dist[AD_matching_vector,] %>% rename(percent_ANI=X10)

matched_AD_mummer_dist$AD_hse <- 1

ctrl_matching_vector <- vector()

for (i in 1:length(ctrl_child_candidates)){
  match1 <- ctrl_child_candidates[i]
  match2 <- ctrl_cg_candidates[i]
  
  row_number_for_subset <- which(str_detect(string=mummer_dist$pattern, 
                                            pattern=match1) & str_detect(string=mummer_dist$pattern, 
                                            pattern=match2))
  
  ctrl_matching_vector <- c(ctrl_matching_vector, row_number_for_subset)
}

matched_ctrl_mummer_dist <- mummer_dist[ctrl_matching_vector,] %>% rename(percent_ANI=X10)

matched_ctrl_mummer_dist$AD_hse <- 0


wilcox.test(x=matched_AD_mummer_dist$percent_ANI, 
            y=matched_ctrl_mummer_dist$percent_ANI,
            alternative="greater")



```

Construct phylogenetic tree using Parsnp output

```{r, parsnp_tree}

parsnp_tree <- read.tree("../data/raw/parsnp_out/parsnp_sub.tree")

#parsnp_tree$tip.label <- gsub(parsnp_tree$tip.label, pattern=".fna", replacement="" )

parsnp_dendrogram <- ggtree(parsnp_tree) + geom_treescale() + geom_tiplab(size=4, color="black")

parsnp_dendrogram 

ggsave("../plots/fig4a_parsnp_tree.pdf", plot=parsnp_dendrogram, width=35, height=12)


```
